src/backend/distributed/master/master_multi_shard_modify.c

```diff
@@ -0,0 +1,336 @@
+
+	List *shardPlacementList = FinalizedShardPlacementList(shardId);
+	ListCell *shardPlacementCell = NULL;
+	List *connectionList = NIL;
+
+	Assert(!shardConnectionsFound);
+
+	if (shardPlacementList == NIL)
+	{
+		ereport(ERROR, (errmsg("Could not find any shard placements for the shard "
+							   UINT64_FORMAT, shardId)));
+	}
+
+	foreach(shardPlacementCell, shardPlacementList)
+	{
+		ShardPlacement *shardPlacement = (ShardPlacement *) lfirst(
+			shardPlacementCell);
+		char *workerName = shardPlacement->nodeName;
+		uint32 workerPort = shardPlacement->nodePort;
+		char *nodeUser = CurrentUserName();
+		PGconn *connection = ConnectToNode(workerName, workerPort, nodeUser);
+		TransactionConnection *transactionConnection = NULL;
+
+		if (connection == NULL)
+		{
+			List *abortConnectionList = ConnectionList(shardConnectionHash);
+			CloseConnections(abortConnectionList);
+
+			ereport(ERROR, (errmsg("Could not establish a connection to all "
+								   "placements.")));
+		}
+
+		transactionConnection = palloc0(sizeof(TransactionConnection));
+
+		transactionConnection->connectionId = shardConnections->shardId;
+		transactionConnection->transactionState = TRANSACTION_STATE_INVALID;
+		transactionConnection->connection = connection;
+
+		connectionList = lappend(connectionList, transactionConnection);
+	}
+
+	shardConnections->connectionList = connectionList;
+}
+
+
+/*
+ * SendQueryToPlacement sends a query to all placement of a shard. Before the actual
+ * query is sent,
+ */
+static int
+SendQueryToConnections(char *shardQueryString, ShardConnections *shardConnections)
+{
+	uint64 shardId = shardConnections->shardId;
+	List *connectionList = shardConnections->connectionList;
+	ListCell *connectionCell = NULL;
+	int32 shardAffectedTupleCount = -1;
+
+	Assert(connectionList != NIL);
+
+	foreach(connectionCell, connectionList)
+	{
+		TransactionConnection *transactionConnection =
+			(TransactionConnection *) lfirst(connectionCell);
+		PGconn *connection = transactionConnection->connection;
+		PGresult *result = NULL;
+		char *placementAffectedTupleString = NULL;
+		int32 placementAffectedTupleCount = -1;
+
+		/* send the query */
+		result = PQexec(connection, "BEGIN");
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			ReportRemoteError(connection, result);
+			ereport(ERROR, (errmsg("Could not send query to shard placement.")));
+		}
+
+		result = PQexec(connection, shardQueryString);
+		if (PQresultStatus(result) != PGRES_COMMAND_OK)
+		{
+			ReportRemoteError(connection, result);
+			ereport(ERROR, (errmsg("Could not send query to shard placement.")));
+		}
+
+		placementAffectedTupleString = PQcmdTuples(result);
+		placementAffectedTupleCount = pg_atoi(placementAffectedTupleString,
+											  sizeof(int32), 0);
+
+		if ((shardAffectedTupleCount == -1) ||
+			(shardAffectedTupleCount == placementAffectedTupleCount))
+		{
+			shardAffectedTupleCount = placementAffectedTupleCount;
+		}
+		else
+		{
+			ereport(WARNING,
```

Comment:
OK, I have changed it to ERROR, instead of WARNING and it causes a ROLLBACK now.